### PR TITLE
refactor(streaming): use LRU cache for group top-n

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As title, the later PR will change cache key from Vec<Datum> to `HashKey`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
close https://github.com/risingwavelabs/risingwave/issues/6484

### CODE REVIEW

File: src/stream/src/executor/top_n/group_top_n.rs [hunk 1]
```diff
@@ -229,12 +268,13 @@ where
...
             // If 'self.caches' does not already have a cache for the current group, create a new
             // cache for it and insert it into `self.caches`
-            if let Vacant(entry) = self.caches.entry(group_key) {
+            if !self.caches.contains(&group_key) {
                 let mut topn_cache = TopNCache::new(self.offset, self.limit, self.order_by_len);
                 self.managed_state
                     .init_topn_cache(Some(&pk_prefix), &mut topn_cache, self.order_by_len)
                     .await?;
-                entry.insert(topn_cache);
+                println!("123");
```
Comment (Little-Wallace) [inline src/stream/src/executor/top_n/group_top_n.rs:276]: Is it a test outout?

Comment (wcy-fdu) [reply src/stream/src/executor/top_n/group_top_n.rs]: Yes, fixed.

**Comment by BugenZhao** [comment]
You should call `evict` manually. 🤣

**Review by st1page — APPROVED** [review]
LGTM

File: src/stream/src/executor/top_n/utils.rs [hunk 2]
```diff
@@ -58,6 +58,9 @@ pub trait TopNExecutorBase: Send + 'static {
         unreachable!()
     }
 
+    fn evict(&mut self) {
+        unreachable!()
+    }
```
Comment (BugenZhao) [inline src/stream/src/executor/top_n/utils.rs:63]:
```suggestion
    }

```

File: src/stream/src/executor/top_n/utils.rs [hunk 3]
```diff
@@ -116,9 +119,10 @@ where
                 Message::Barrier(barrier) => {
                     self.inner.flush_data(barrier.epoch).await?;
 
-                    // Update the vnode bitmap, only used by Group Top-N.
+                    // Update the vnode bitmap and evict cache, only used by Group Top-N.
                     if let Some(vnode_bitmap) = barrier.as_update_vnode_bitmap(self.ctx.id) {
                         self.inner.update_vnode_bitmap(vnode_bitmap);
+                        self.inner.evict();
```
Comment (BugenZhao) [inline src/stream/src/executor/top_n/utils.rs:125]: The eviction is irrelevant to the scaling. We should always call the `evict`.

**Review by BugenZhao — CHANGES_REQUESTED** [review]

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**